### DISCUSSION TITLE
Tasks and Priorities

### DISCUSSION THREAD

**Discussion by ColtAllen** [ColtAllen/btyd#47] [General]
## This post will be actively maintained, so keep checking back and if there are any specific tasks or suggestions you want to work on for the Beta release and beyond, please let me know.

[Rodrigorivera](https://github.com/ColtAllen/lifetimes/issues/14) inspired me to create [Projects ](https://github.com/ColtAllen/lifetimes/projects?type=beta) for all outstanding tasks, which will be sorted by version release:

<details><summary><b>BETA RELEASE</b></summary>
<p>

**[Deprecate `fitters` module](https://github.com/ColtAllen/lifetimes/discussions/18) and fix known `lifetimes` bugs:**

- [Create Aesara `Op` for `scipy.special.hyp2f1`](https://github.com/ColtAllen/btyd/issues/49)
- [Investigate Monetary Value Calculations](https://github.com/ColtAllen/btyd/issues/51)
- Add `pareto_nbd_model.py` module
- Add `bg_bb_model.py` module
- Add `bg_nbd_covars_model.py` module
- Raise `pymc` exception whenever model fails to converge
- [Fix Plotting Dependencies on Calibration vs. Holdout Purchases](https://github.com/ColtAllen/btyd/issues/50)
- Add matrix plotting method to `BaseModel`
- Type hinting for `utils.py`, `plotting.py`, and `generate_data.py` modules

</p>
</details>

<details><summary><b>RELEASE CANDIDATE</b></summary>
<p>

**Polish up documentation and fix any bugs identified while working on beta release(s):**

- Add `arviz` plotting suggestions and code to docs for model inference
- Add an explanatory chart to documentation summarizing model characteristics
- Create and add logo for library
- Add troubleshooting guide to documentation
- Resolve documentation theme
- Docstrings for predictive methods.
- CI/CD for auto-updates to project documentation
- Consider hosting docs on GitHub and emulating documentation style of ArViZ and `bambi`

</p>
</details>

<details><summary><b>FUTURE VERSION RELEASES</b></summary>
<p>

**Implement additional BTYD Models, enhance plotting functions, documentation and library performance, etc.**

- Add time-invariant covariates to `ParetoNBDModel`
- Add RFM segmentation utility
- Add Pareto/GGG Model
- Add PDO Model
- Add test(s) comparing predictions between hourly and daily aggregations.
- Add BG/NBD-k and MBG/NBD-k Models
- Add time-varying covariates to `BetaGeoModel`
- Add time-varying covariates to `ParetoNBDModel`? (Note this model could take days to train and necessitate checkpointing functionality to be added)
- Suggestions for any other models to add?
- Add tutorial notebook(s)
- Add sample_kwargs to `BaseModel.fit`
- Fix BetaGeo conditional probability calculation
- Review viability of modifying hyperprior RVs
- Enhance existing plotting functions with `seaborn`
- `jax` sampling via `pymc` to speed up model fitting times
- Comprehensive evaluation of model priors to determine best defaults
-  [Refactor `utils.py`](https://github.com/ColtAllen/lifetimes/discussions/19)
- Add code for `mlflow` model registration to documentation
- Refactor `BetaGeoModel._conditional_probability_alive` unit test to remove extraneous args.
- Refactor `BetaGeoModel._conditional_expected_number_of_purchases_up_to_time` unit test to remove extraneous args.
- Delete redundant `load_cdnow_summary()` function
- Try `GammaGammaModel._log_likelihood` variant

</p>
</details>

## Projects

* **Documentation Adventures** [(Discussion)](https://github.com/ColtAllen/lifetimes/discussions/17)
* **Deprecate `fitters` Module** [(Discussion)](https://github.com/ColtAllen/lifetimes/discussions/18)
* **Refactor `utils.py`** [(Discussion)](https://github.com/ColtAllen/lifetimes/discussions/19)
* **Plotting Parea** [(Discussion)](https://github.com/ColtAllen/lifetimes/discussions/20)
* **New Models to Add** [(Discussion)](https://github.com/ColtAllen/lifetimes/discussions/21)
* **CI/CD** [(Discussion)](https://github.com/ColtAllen/lifetimes/discussions/22)
* **Library Optimization** [(Discussion)](https://github.com/ColtAllen/lifetimes/discussions/23)

**Comment by clintdunn** [comment]
@ColtAllen Do you have more context on each of the tasks? I'd like to investigate a few but not sure what needs to be done

**Comment by ColtAllen** [comment]
@clintdunn I appreciate the early enthusiasm 👍 I've been writing descriptions for each of the above tasks. After the alpha release is published I'll post them all as issues for collaborators to review and work on.

I'm targeting this Fri, 17-Jun for the alpha release (and task/issue postings) because I'm going on vacation afterward and won't be able to resume work on `lifetimes` until July.

**Comment by clintdunn** [comment]
@ColtAllen sounds good!

**Comment by ColtAllen** [comment]
**EDIT: I'm back from vacation now and will be working on all alpha release issues.**

Just posted two issues if anyone wants to work on them 👍 I consider posterior parameter sampling for predictions  to be the most important task, but it would be best if the issues I just posted were completed first as it will simplify the work required.

**Also, if anyone wants me to plan a Zoom call next month to discuss ongoing and future work, let me know.**

**Comment by ColtAllen** [comment]
I'm back from vacation and have reorganized work tasks by version release. I'll handle everything for the alpha release, but there are a lot of fairly straightforward tasks to work on for the beta I could use some help with, nearly all of which involve modules that will be left as-is while in alpha. More information about these tasks will be provided in the pertinent project discussions and posted as issues soon.

**Comment by ColtAllen** [comment]
First version of Beta release is now available for `pip` installation! I intend to publish new Beta releases as new models are added.

**Reply by zwelitunyiswa** [reply]
@ColtAllen This is great. I have been waiting for a bayesian CLV in python (a la BTYD Plus in R). This is fantastic. Count me in for contributing documentation and the like. I have no experience in Github, but this would be a great opportunity to learn.

**Reply by ColtAllen** [reply]
Thanks @zwelitunyiswa! [The `bambi` contribution guide](https://github.com/bambinos/bambi/blob/main/CONTRIBUTING.md) is a great overview on how to use GitHub for open source projects, and may inspire some future additions to this library's CONTRIBUTING.md file.

The documentation could benefit from some refinements before the official release is published, so any feedback you have on the layout, color scheme, ideas to add, etc. would be helpful. Currently the documentation uses the _pydata_sphinx_theme_, but I'm not opposed to returning to the _sphinx_rtd_theme_ used in the old documentation. The [Kedro documentation](https://kedro.readthedocs.io/en/stable/index.html) uses that same theme in fact, albeit heavily customized.

**Comment by ColtAllen** [comment]
Second version of Beta release is now available for pip installation! This one includes a Bayesian variant of the Gamma-Gamma model and a few bug fixes.

**Comment by ColtAllen** [comment]
Third version of Beta release is now available for pip installation! This one includes a Bayesian variant of the Modified BG/NBD model, a few bug fixes, and requested enhancements to the legacy `fitters.py` package.
